### PR TITLE
Alert user before closing the createItemModal form

### PR DESCRIPTION
**Pull request by singhArmani**
The form data inside the `createItemModal` was getting lost when user
hit the `cancel` button. This could lead to a poor UX.

This fix will identifies if form data has changed from its initial
values before closing the modal. If yes, then it alerts the user for
confirmation. Otherwise, will close the modal.
admin

Closes: #2822

### CODE REVIEW

File: packages/app-admin-ui/client/components/CreateItemModal.js [hunk 1]
```diff
@@ -244,10 +271,36 @@ const CreateItemModal = ({ prefillData = {}, onClose, onCreate, viewOnSave }) =>
               ));
             }}
           </Render>
+          <ConfirmModal
+            isOpen={isConfirmOpen}
+            onYes={() => {
```
Comment (jordanoverbye) [inline packages/app-admin-ui/client/components/CreateItemModal.js:276]: I'd call this `onConfirm` and `onCancel`

File: packages/app-admin-ui/client/components/CreateItemModal.js [hunk 2]
```diff
@@ -129,16 +131,41 @@ const CreateItemModal = ({ prefillData = {}, onClose, onCreate, viewOnSave }) =>
...
     closeCreateItemModal();
     setItem(list.getInitialItemData({}));
-    const data = arrayToObject(creatable, 'path', field => field.serialize(item));
     if (onClose) {
+      const data = arrayToObject(creatable, 'path', field => field.serialize(item));
       onClose(data);
     }
   };
 
+  const _onClose = () => {
```
Comment (jordanoverbye) [inline packages/app-admin-ui/client/components/CreateItemModal.js:159]: Can you make this use `useCallback`

File: packages/app-admin-ui/client/components/CreateItemModal.js [hunk 3]
```diff
@@ -129,16 +131,41 @@ const CreateItemModal = ({ prefillData = {}, onClose, onCreate, viewOnSave }) =>
...
+    for (const path of Object.keys(currentValues)) {
+      if (data[path].hasChanged(initialValues, currentValues)) {
+        hasChanged = true;
+        break;
+      }
+    }
+    return hasChanged;
+  };
+
+  const _createItemModalClose = () => {
```
Comment (jordanoverbye) [inline packages/app-admin-ui/client/components/CreateItemModal.js:150]: Can you make this use `useCallback`

File: packages/app-admin-ui/client/components/CreateItemModal.js [hunk 4]
```diff
@@ -129,16 +131,41 @@ const CreateItemModal = ({ prefillData = {}, onClose, onCreate, viewOnSave }) =>
     }
   });
 
-  const _onClose = () => {
-    if (loading) return;
+  // Identifies if the user has changed the initial data in the form.
+  const hasFormDataChanged = () => {
```
Comment (jordanoverbye) [inline packages/app-admin-ui/client/components/CreateItemModal.js:135]: Can you make this use `useCallback`

File: packages/app-admin-ui/client/components/CreateItemModal.js (hunk 1 above)
Comment (jordanoverbye) [reply packages/app-admin-ui/client/components/CreateItemModal.js:276]: I'd also move these inline functions to callbacks too.

```
 <ConfirmModal
  isOpen={isConfirmOpen}
  onConfirm={onConfirm}
  onCancel={onCancel}
/>
```

**Comment by MadeByMike** [comment]
Looks pretty good. I do agree about changing `onYes` to `onConfirm`. 

I also wondering if the has data changed logic could be more simple? Normally I'd just set an `isDirty` flag when someone interacts with the form elements. It's not perfect but avoids owning more logic. -- What you've done looks ok. This is just a question.

Make what changes you think necessary and I'll ✅

File: packages/app-admin-ui/client/components/CreateItemModal.js (hunk 4 above)
Comment (singhArmani) [reply packages/app-admin-ui/client/components/CreateItemModal.js:135]: This may not require useCallback decorator as it's not being passed as a prop.

**Comment by singhArmani** [comment]
> This is looking good! I just have some comments around variable names and callbacks.

I have refactored the variables' names 👍 . But I am not sure I agree with wrapping all the functions with `useCallback` hook. `useCallback` does have some cost associated with it. And IMO seems pre-optimization.

**Comment by MadeByMike** [comment]
> I am not sure I agree with wrapping all the functions with `useCallback` hook.  

I care little about the pre-optimisation cost or value of the hook (in this specific case) - all I care about here is how readable the code is and I think there's little difference either way here. Not going to hold out for that change.

File: packages/app-admin-ui/client/components/CreateItemModal.js [hunk 5]
```diff
@@ -147,6 +174,17 @@ const CreateItemModal = ({ prefillData = {}, onClose, onCreate, viewOnSave }) =>
...
+  const _onConfirm = () => {
+    setConfirmOpen(false);
+    _createItemModalClose();
+  };
+
+  // Handler when user hits the cancel on confirm modal
+  const _onCancel = () => {
+    setConfirmOpen(false);
+  };
+
```
Comment (MadeByMike) [inline packages/app-admin-ui/client/components/CreateItemModal.js:187]:
```suggestion
```

File: packages/app-admin-ui/client/components/CreateItemModal.js [hunk 6]
```diff
@@ -244,10 +282,29 @@ const CreateItemModal = ({ prefillData = {}, onClose, onCreate, viewOnSave }) =>
               ));
             }}
           </Render>
+          <ConfirmModal isOpen={isConfirmOpen} onConfirm={_onConfirm} onCancel={_onCancel} />
```
Comment (MadeByMike) [inline packages/app-admin-ui/client/components/CreateItemModal.js:285]:
```suggestion
          <ConfirmModal 
            isOpen={isConfirmOpen} 
            onConfirm={() => {
              setConfirmOpen(false);
              _createItemModalClose();
            }} 
            onCancel={() => {
              setConfirmOpen(false);
            }} 
          />
```

Comment (MadeByMike) [reply packages/app-admin-ui/client/components/CreateItemModal.js:285]: You might need to run prettier after this.

File: packages/app-admin-ui/client/components/CreateItemModal.js (hunk 1 above)
Comment (MadeByMike) [reply packages/app-admin-ui/client/components/CreateItemModal.js]: Yeah that was exactly my comment :)

Comment (MadeByMike) [reply packages/app-admin-ui/client/components/CreateItemModal.js]: What's the value in moving an inline function unless it's used more than once or is huge? 

My preference is usually to keep the code as small as possible until it hinders readability.

**Comment by singhArmani** [comment]
> Normally I'd just set an `isDirty` flag when someone interacts with the form elements. It's not perfect but avoids owning more logic. -- What you've done looks ok. This is just a question.

I completely agree that the logic is verbose, and should have a boolean flag to check the form state. But implementing `isDirty` may be tricky too IMO. It's not only that if the user has interacted (touched),  but also if the form values have changed from its previous values.

**Pull request merged**